### ISSUE TITLE
[Feature Request] Sparse lazy tensor for SKI wights 

### ISSUE THREAD

**Issue by ymohit** [cornellius-gp/gpytorch#1680]
# 🚀 Feature Request

Currently, SKI weights (typically, written as W) are handled through two matrices containing indices and values separately. Also, special routines are written in utils to handle MVMs with them. These two things make it suitable for a new lazy tensor. 

## Motivation
As a result of this feature, the SKI code will be more readable. 

**Are you willing to open a pull request?** (We LOVE contributions!!!)
I'm happy to give it a try but as of now I'm not sure about two things: a) if at all such an abstraction will be computationally efficient (or even more efficient than utils.sparse.bdsmm procedure)? b) I'm not able to understand how batch shape is organized in this project and in particular in these functions. Would it be possible to have a sparse SKI weight in the batch made if such a lazy tensor is created?

**Comment by wjmaddox** [comment]
Hi, I recently did a bunch of work using the matrices `W` in SKI, and this would be pretty helpful. For GPs, I think we generally consider solely the type of operations of `W K W' v` which is served by `InterpolatedLazyTensor` but for online learning with GPs, I know that one might want better access to `W` (e.g. `W v` or `W'W v`) instead...

Responding directly to a) if done properly, it should be give or take as computationally efficient as the current set of operations (maybe a bit slower due to slightly more overhead, outweighed by potential downstream usefulness) and b) lazy tensors should support batch mode when implemented correctly, in general, all up to the last dimension is considered batch dimensions. I could help support an attempt for constructing a SparseLazyTensor.

To develop a `fantasy_strategy` (e.g. conditioning a SKI model on new inputs), I ended up actually densifying the `W` matrices [here](https://github.com/cornellius-gp/gpytorch/blob/b237019940620b1f8546a28b71b632599684769f/gpytorch/models/exact_prediction_strategies.py#L374), but a lazy tensor would probably also have been useful :).

Re your second question, we need gradients of values for Bayesian optimization where one typically uses gradient based optimization with respect to input data points at test time (or even occasionally the training data itself). See [botorch repo](https://github.com/pytorch/botorch).

**Comment by ymohit** [comment]
Thanks for responding @wjmaddox. I believe even W'W is likely to be sparse in many scenarios (e.g., https://arxiv.org/abs/2101.11751) which further suggests we could have lazy tensor even for that. 

Help on support SparseLazyTensor would be super useful. Thanks a lot!

Thanks for the pointer on Bayesian optimization though I still don't quite understand the details of gradient wrt values is utilized. Can you refer for a paper on Bayesian that provide details on how SKI is utilized?

**Comment by wjmaddox** [comment]
Ahh yeah, thanks for the pointer to your paper --- we had a similar [paper](https://arxiv.org/abs/2103.01454) (also at aistats) where we were taking a similar approach for online learning with ski . We also did a bunch of bayes opt in that setting. 

But in our experience, sparse cholesky factorizations weren't particularly memory efficient for fewer than 5000 or so inducing points. Would be happy to revisit.

See that paper for details on the bayes opt settings, and the experimental code is [here](https://github.com/wjmaddox/online_gp/tree/main/experiments/bayesopt).

**Comment by ymohit** [comment]
@wjmaddox I've created a corresponding PR to complete this feature request and mentioned steps based on what I've understood from our conversation here. Please feel free to add and make suggestions if you have any. I'd appreciate help in creating the first step. I'm happy to push one dummy sparse lazy tensor without batch dimensions though. Would that make sense?

**Comment by wjmaddox** [comment]
Sorry, missed this comment, go ahead and try to push the sparse lazy tensor without batch dimensions and I can probably get it working in batch mode. 

I also managed to track down a couple of structured lazy tensors (`BandedSparseLazyTensor` and `AsymBandedSparseLazyTensor`) that we (iirc @idelbrid) had written to account for `W'W` and can add them in to your PR after you put up the sparse LT.

**Comment by ymohit** [comment]
Thanks, @wjmaddox. It will be delightful to add these two structured lazy tensors into the PR. 

I've tried to push one initial version of the sparse lazy tensor based on my current understanding including batch operations otherwise it'd be just 'torch.sparse tensor'.  Though, I'm not super clear now about how to write tests. 

a) Can you please suggest what are essential tests for the lazy tensor to ensure batch operations are correct? Are they all included in LazyTensorTestCase? Disclaimer: I've done no testing/debugging of the current code. 

b) Clarification on _matmul: assuming b1 x ... x bk X m x n be the sparse matrix tensor size, what are the dimensions of rhs should be in def _matmul(self, rhs: Tensor) -> Tensor? Should it be some n X p or b1 x ... x bk X n X p or  b_new X n X p? Sorry, I'm new to gpytorch formatting. 

c) I'm convinced of the sheer simplification of sparse_lazy_tensor to make the case for the PR. I'm wondering what would be a proper performance test that could respond to Jacob's apprehension about the efficiency of sparse_lazy_tensor.  Is a comparison of MVMs with SKI W matrix on both cpu and gpu will be good enough?

**Comment by gpleiss** [comment]
> I'm wondering what would be a proper performance test that could respond to Jacob's apprehension about the efficiency of sparse_lazy_tensor. Is a comparison of MVMs with SKI W matrix on both cpu and gpu will be good enough?

Yeah a rough performance comparison on cpu and gpu would be great!
